C compilation error while matching generic type

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 1b9f15d, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 1b9f15d
|:-------------------|:-------------------
|OS                  |macos, macOS, 14.4, 23E214
|Processor           |12 cpus, 64bit, little endian, Apple M2 Pro
|                    |
|V last modified time|2025-01-18 06:25:43
|                    |
|Git version         |git version 2.39.3 (Apple Git-145)
|V git status        |1b9f15d6
|.git/config present |true
|                    |
|cc version          |Apple clang version 15.0.0 (clang-1500.1.0.2.5)
|gcc version         |Apple clang version 15.0.0 (clang-1500.1.0.2.5)
|clang version       |Apple clang version 15.0.0 (clang-1500.1.0.2.5)
|tcc version         |tcc version 0.9.28rc 2024-02-05 HEAD@105d70f7 (AArch64 Darwin)
|tcc git status      |thirdparty-macos-arm64 c8df4e27
|emcc version        |N/A
|glibc version       |N/A

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  ./src/main.v && ./src/main`
```v
module main

import strconv

fn main() {
	println(to_int('1'))
	println(to_int(1))
}

fn to_int[T](v T) i64 {
	return match typeof(v).name {
		'string' {
			strconv.atoi(v) or { 0 }
		}
		'int' {
			v
		}
		else {
			0
		}
	}
}
```

**What did you see?**
```
================== C compilation error (from cc): ==============
cc: /tmp/v_501/main.01JHW86QMJ2K76ZPQNW05P61W8.tmp.c:5804:7: error: assigning to 'int' from incompatible type 'string' (aka 'struct string')
cc:                 _t2 = v;
cc:                     ^ ~
cc: /tmp/v_501/main.01JHW86QMJ2K76ZPQNW05P61W8.tmp.c:5814:33: error: passing 'int' to parameter of incompatible type 'string' (aka 'struct string')
cc: _result_int _t3 = strconv__atoi(v);
cc:                                 ^
cc: /tmp/v_501/main.01JHW86QMJ2K76ZPQNW05P61W8.tmp.c:1906:34: note: passing argument to parameter 's' here
cc: _result_int strconv__atoi(string s) {
cc:                                  ^
cc: 2 errors generated.
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

no errors, as string should never take the int branch and vice versa

**Equivalent Go code**
```go
package main

import (
	"fmt"
	"strconv"
)

func main() {
	fmt.Println(to_int("1"))
	fmt.Println(to_int(1))
}

func to_int[T any](v T) int64 {
	switch v := any(v).(type) {
	case string:
		i, err := strconv.Atoi(v)
		if err != nil {
			return 0
		}
		return int64(i)
	case int:
		return int64(v)
	default:
		return 0
	}
}
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/match.v
+++ b/vlib/v/checker/match.v
@@ -132,13 +132,13 @@ fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 						stmt_sym := c.table.sym(stmt.typ)
 						if ret_sym.kind !in [.sum_type, .interface]
 							&& stmt_sym.kind in [.sum_type, .interface] {
-							c.error('return type mismatch, it should be `${ret_sym.name}`',
+							c.error('return type mismatch, it should be `${ret_sym.name}`, but it is instead `${c.table.type_to_str(expr_type)}`',
 								stmt.pos)
 						}
 						if ret_type.nr_muls() != stmt.typ.nr_muls()
 							&& stmt.typ.idx() !in [ast.voidptr_type_idx, ast.nil_type_idx] {
 							type_name := '&'.repeat(ret_type.nr_muls()) + ret_sym.name
-							c.error('return type mismatch, it should be `${type_name}`',
+							c.error('return type mismatch, it should be `${type_name}`, but it is instead `${c.table.type_to_str(expr_type)}`',
 								stmt.pos)
 						}
 					}
@@ -281,7 +281,8 @@ fn (mut c Checker) check_match_branch_last_stmt(last_stmt ast.ExprStmt, ret_type
 					return
 				}
 			}
-			c.error('return type mismatch, it should be `${ret_sym.name}`', last_stmt.pos)
+			c.error('return type mismatch, it should be `${ret_sym.name}`, but it is instead `${c.table.type_to_str(expr_type)}`',
+				last_stmt.pos)
 		}
 	} else if expr_type == ast.void_type && ret_type.idx() == ast.void_type_idx
 		&& ret_type.has_option_or_result() {

--- a/vlib/v/checker/match.v
+++ b/vlib/v/checker/match.v
@@ -76,11 +76,13 @@ fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 				} else {
 					c.expr(mut stmt.expr)
 				})
+				unwrapped_expected_type := c.unwrap_generic(node.expected_type)
 				stmt.typ = expr_type
 				if first_iteration {
-					if node.expected_type.has_option_or_result()
-						|| c.table.type_kind(node.expected_type) in [.sum_type, .multi_return] {
-						c.check_match_branch_last_stmt(stmt, node.expected_type, expr_type)
+					if unwrapped_expected_type.has_option_or_result()
+						|| c.table.type_kind(unwrapped_expected_type) in [.sum_type, .multi_return] {
+						c.check_match_branch_last_stmt(stmt, unwrapped_expected_type,
+							expr_type)
 						ret_type = node.expected_type
 					} else {
 						ret_type = expr_type
@@ -105,10 +107,9 @@ fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 					}
 				} else {
 					if ret_type.idx() != expr_type.idx() {
-						if node.expected_type.has_option_or_result()
+						if unwrapped_expected_type.has_option_or_result()
 							&& c.table.sym(stmt.typ).kind == .struct
-							&& (c.table.sym(ret_type).kind != .sum_type
-							|| !c.check_types(expr_type, ret_type))
+							&& !c.check_types(expr_type, c.unwrap_generic(ret_type))
 							&& c.type_implements(stmt.typ, ast.error_type, node.pos) {
 							stmt.expr = ast.CastExpr{
 								expr:      stmt.expr
@@ -119,7 +120,8 @@ fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 							}
 							stmt.typ = ast.error_type
 						} else {
-							c.check_match_branch_last_stmt(stmt, ret_type, expr_type)
+							c.check_match_branch_last_stmt(stmt, c.unwrap_generic(ret_type),
+								expr_type)
 							if ret_type.is_number() && expr_type.is_number() && !c.inside_return {
 								ret_type = c.promote_num(ret_type, expr_type)
 							}

--- a/vlib/v/checker/match.v
+++ b/vlib/v/checker/match.v
@@ -71,11 +71,11 @@ fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 				} else {
 					node.expected_type
 				}
-				expr_type := if stmt.expr is ast.CallExpr {
+				expr_type := c.unwrap_generic(if stmt.expr is ast.CallExpr {
 					stmt.typ
 				} else {
 					c.expr(mut stmt.expr)
-				}
+				})
 				stmt.typ = expr_type
 				if first_iteration {
 					if node.expected_type.has_option_or_result()

--- a/vlib/v/checker/tests/match_generic_case_err.out
+++ b/vlib/v/checker/tests/match_generic_case_err.out
@@ -1,4 +1,4 @@
-vlib/v/checker/tests/match_generic_case_err.vv:16:4: error: return type mismatch, it should be `int`
+vlib/v/checker/tests/match_generic_case_err.vv:16:4: error: return type mismatch, it should be `int`, but it is instead `string`
    14 |         }
    15 |         'int' {
    16 |             v

--- a/vlib/v/checker/tests/match_generic_case_err.out
+++ b/vlib/v/checker/tests/match_generic_case_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/match_generic_case_err.vv:16:4: error: return type mismatch, it should be `int`
+   14 |         }
+   15 |         'int' {
+   16 |             v
+      |             ^
+   17 |         }
+   18 |         else {

--- a/vlib/v/checker/tests/match_generic_case_err.vv
+++ b/vlib/v/checker/tests/match_generic_case_err.vv
@@ -1,0 +1,22 @@
+module main
+
+import strconv
+
+fn main() {
+	println(to_int('1'))
+	println(to_int(1))
+}
+
+fn to_int[T](v T) i64 {
+	return match typeof(v).name {
+		'string' {
+			strconv.atoi(v) or { 0 }
+		}
+		'int' {
+			v
+		}
+		else {
+			0
+		}
+	}
+}

--- a/vlib/v/checker/tests/match_return_mismatch_type_err.out
+++ b/vlib/v/checker/tests/match_return_mismatch_type_err.out
@@ -5,35 +5,35 @@ vlib/v/checker/tests/match_return_mismatch_type_err.vv:27:6: warning: cannot ass
       |         ^
    28 |         string { &any }
    29 |         else { &variable }
-vlib/v/checker/tests/match_return_mismatch_type_err.vv:4:10: error: return type mismatch, it should be `string`
+vlib/v/checker/tests/match_return_mismatch_type_err.vv:4:10: error: return type mismatch, it should be `string`, but it is instead `int literal`
     2 |     a := match 1 {
     3 |         1 { 'aa' }
     4 |         else { 22 }
       |                ~~
     5 |     }
     6 |     println(a)
-vlib/v/checker/tests/match_return_mismatch_type_err.vv:18:10: error: return type mismatch, it should be `&string`
+vlib/v/checker/tests/match_return_mismatch_type_err.vv:18:10: error: return type mismatch, it should be `&string`, but it is instead `string`
    16 |     _ = match any {
    17 |         string { &any }
    18 |         else { variable }
       |                ~~~~~~~~
    19 |     }
    20 |
-vlib/v/checker/tests/match_return_mismatch_type_err.vv:23:10: error: return type mismatch, it should be `string`
+vlib/v/checker/tests/match_return_mismatch_type_err.vv:23:10: error: return type mismatch, it should be `string`, but it is instead `&string`
    21 |     _ = match any {
    22 |         string { any }
    23 |         else { &variable }
       |                ^
    24 |     }
    25 |
-vlib/v/checker/tests/match_return_mismatch_type_err.vv:43:10: error: return type mismatch, it should be `&string`
+vlib/v/checker/tests/match_return_mismatch_type_err.vv:43:10: error: return type mismatch, it should be `&string`, but it is instead `string`
    41 |     _ = match any {
    42 |         string { &any }
    43 |         else { variable }
       |                ~~~~~~~~
    44 |     }
    45 |
-vlib/v/checker/tests/match_return_mismatch_type_err.vv:48:10: error: return type mismatch, it should be `string`
+vlib/v/checker/tests/match_return_mismatch_type_err.vv:48:10: error: return type mismatch, it should be `string`, but it is instead `&string`
    46 |     _ = match any {
    47 |         string { any }
    48 |         else { &variable }

--- a/vlib/v/checker/tests/match_return_sumtype_mismatch_err.out
+++ b/vlib/v/checker/tests/match_return_sumtype_mismatch_err.out
@@ -1,4 +1,4 @@
-vlib/v/checker/tests/match_return_sumtype_mismatch_err.vv:15:11: error: return type mismatch, it should be `Myt`
+vlib/v/checker/tests/match_return_sumtype_mismatch_err.vv:15:11: error: return type mismatch, it should be `Myt`, but it is instead `rune`
    13 |     return match b {
    14 |         true { St('TRUE') }
    15 |         false { `F` }

--- a/vlib/v/checker/tests/return_match_expr_type_mismatch.out
+++ b/vlib/v/checker/tests/return_match_expr_type_mismatch.out
@@ -1,4 +1,4 @@
-vlib/v/checker/tests/return_match_expr_type_mismatch.vv:18:16: error: return type mismatch, it should be `SomeTuple`
+vlib/v/checker/tests/return_match_expr_type_mismatch.vv:18:16: error: return type mismatch, it should be `SomeTuple`, but it is instead `Poss1`
    16 | fn get_file(item PossOwner) ?SomeTuple {
    17 |     return match item.pos {
    18 |         Poss1 { item.pos }

--- a/vlib/v/tests/generics/generic_match_expr_test.v
+++ b/vlib/v/tests/generics/generic_match_expr_test.v
@@ -1,0 +1,24 @@
+struct Foo[T] {
+	a T
+}
+
+fn r[T]() Foo[T] {
+	return Foo[T]{}
+}
+
+fn t[T](v T) !Foo[T] {
+	return match typeof(v).name {
+		'string' {
+			r[T]()
+		}
+		else {
+			r[T]()
+		}
+	}
+}
+
+fn test_main() {
+	t(1)!
+	t('')!
+	assert true
+}

